library/alloc/src/sync.rs

```diff
@@ -943,11 +949,16 @@ impl<T: ?Sized> Arc<T> {
         let mut cur = this.inner().weak.load(Relaxed);
 
         loop {
-            // check if the weak counter is currently "locked"; if so, spin.
-            if cur == usize::MAX {
-                hint::spin_loop();
-                cur = this.inner().weak.load(Relaxed);
-                continue;
+            if cur > MAX_REFCOUNT {
+                // check if the weak counter is currently "locked"; if so, spin.
+                if cur == usize::MAX {
+                    hint::spin_loop();
+                    cur = this.inner().weak.load(Relaxed);
+                    continue;
+                }
+                // We shouldn't allow the refcount to increase much past `MAX_REFCOUNT`.
+                // See https://github.com/rust-lang/rust/issues/108706
+                panic!("{}", INTERNAL_OVERFLOW_ERROR);
```

Comment:
We can't panic here, because panics can be caught with `panic=unwind`. This needs to be an abort.

Reply:
It's okay if it gets caught, because the counter hasn't been incremented yet.
As long as it hasn't been incremented, the program can continue.

To the best of my understanding, it can only happen in already-buggy code, which is why aborting in an undocumented way is considered fine. So this wouldn't supposedly be a breaking change.

I figured panicking would be nicer than aborting, and change the `upgrade` method to match as well.

Reply:
(Since this is in a compare-exchange loop instead of a `fetch_add` or similar).